src/game_events/action_wml.cpp

```diff
@@ -239,20 +240,40 @@ wml_action::wml_action(const std::string & tag, handler function)
  *
  * Generated code looks like this:
  * \code
- * void wml_func_foo(...);
+ * static void wml_func_foo(...);
  * static wml_action wml_action_foo("foo", &wml_func_foo);
- * void wml_func_foo(...)
+ * static void wml_func_foo(...)
  * {
  *    // code for foo
  * }
  * \endcode
  */
 #define WML_HANDLER_FUNCTION(pname, pei, pcfg) \
-	static void wml_func_##pname(const queued_event &pei, const vconfig &pcfg); \
+	static void wml_func_##pname(const queued_event& pei, const vconfig& pcfg); \
 	static wml_action wml_action_##pname(#pname, &wml_func_##pname);  \
 	static void wml_func_##pname(const queued_event& pei, const vconfig& pcfg)
 
 
+WML_HANDLER_FUNCTION(set_achieved,,pcfg)
```

Comment:
Ohhh? A new C++-implemented WML tag? I believe we ultimately would prefer to phase out this entire file, so I'm not sure if it's a good idea to add new stuff here. It might be better to add a Lua API function and then implement the tag in Lua.

Reply:
What I wasn't sure about was where to put it - there didn't seem to be any good module to add the C++ part of the implementation to other than maybe `wml-utils`.

Reply:
You said C++ but I think you meant Lua?

There's certainly no module for achievements. I think something like `wesnoth.achievements` would be the best place for this API. I think it might be nice if it can be used like this:

```lua
function wesnoth.wml_actions.set_achieved(cfg)
  -- Maybe like this?
  wesnoth.achievements[cfg.content_for][cfg.id] = true
  -- Or like this?
  wesnoth.achievements[cfg.content_for].achieve(cfg.id)
end

function wesnoth.wml_actions.has_achievement(cfg)
  -- Maybe like this?
  return wesnoth.achievements[cfg.content_for][cfg.id]
  -- Or like this?
  wesnoth.achievements[cfg.content_for].has(cfg.id)
end
```

As for what files those functions would go in, normally wml-tags.lua and wml-conditionals.lua would be the correct place, unless they were long, complicated functions, in which case they'd go in wml/tag-name.lua.

Reply:
No, I meant C++. This can't be implemented in pure lua due to needing to write newly completed achievements to the preferences file, unless there's already an API for that. `has_achievement` also wouldn't go under wml_actions, would it?

Reply:
> there didn't seem to be any good module to add the C++ part of the implementation to

Then, I have no idea what you mean by this. There's basically no code in the WML tag implementation.

> `has_achievement` also wouldn't go under wml_actions, would it?

Yes, that was my mistake, sorry. I corrected it.

Reply:
> Then, I have no idea what you mean by this. There's basically no code in the WML tag implementation.

I mean that I need to write new C++ code *somewhere* to do the preferences reading and writing for these tags, don't I?

Reply:
Well, there is basically no code in the existing functions for those tags. It's literally taking the input from the tag and passing it on to another function. So yeah, moving that into game_lua_kernel.cpp would be perfectly fine.

Reply:
I've added it instead as `wesnoth.achievements.get()` and `wesnoth.achievements.has()`. I'm not really sure how to have it use the square brackets for access - is that what `__index` and `__newindex` are for?

Reply:
Yes. There are examples of it being done for other APIs. Using `get` and `has` isn't a terrible API though, and if you think there's some reason to possibly add other functions in the future, there might be some argument that it's better. Personally, barring someone thinking of such a potential future API, I still prefer square-bracket access.

The only other function I can think of would be a way to iterate through the available achievements (or is there some reason that's a bad idea?). Of course, it _would_ be nice if that could just be done with `pairs`…

Reply:
I think I'll leave it as `get()` and `has()`. There's the additional functionality of having "in progress" achievements that's been mentioned and it wouldn't surprise me at all if people came up with more things to do with achievements.

Reply:
I could imagine "in progress" achievements to be distinguished as being int-valued instead of bool-valued if using the bracket syntax, but I don't think I mind leaving it as it is.

But, is there any reason not to add a third function that lists all available achievements? (Not sure how that function would handle hidden achievements. It could exclude them, include them but label them as hidden, include them only if the user has them unlocked, or take an input to specify whether to include them, which could be either a boolean true/false or an enum string selecting one of the first three options I listed here.)

Reply:
Mainly just that I couldn't think of a use case for where it would be useful.